text/0000-stable-simd.md

```diff
@@ -0,0 +1,718 @@
+names a proposed for becoming stable in the first pass.
+
+## The `target_feature` value in `#[cfg]`
+
+In addition to enabling target features for a function the compiler will also
+allow statically testing whether a particular target feature is enabled. This
+corresponds to the `cfg_target_feature` feature today in rustc, and can be seen
+via:
+
+```rust
+#[cfg(target_feature = "avx)]
+fn foo() {
+    // implementation that can use `avx`
+}
+
+#[cfg(not(target_feature = "avx))]
+fn foo() {
+    // a fallback implementation
+}
+```
+
+Additionally this is also made available to `cfg!`:
+
+```rust
+if cfg!(target_feature = "avx") {
+    println!("this program was compiled with AVX support");
+}
+```
+
+The `#[cfg]` attribute and `cfg!` macro statically resolve and **do not do
+runtime dispatch**. Tweaking these functions is currently done via the `-C
+target-feature` flag to the compiler. This flag to the compiler accepts a
+similar set of strings to the ones specified above and is already "stable".
+
+## The `cfg_feature_enabled!` Macro
+
+One mode of operation with vendor intrinsics is to compile *part* of a program
+with certain CPU features enabled but not the entire program. This way a
+portable program can be compiled which runs across a broad range of hardware
+which can still benefit from optimized implementations for particular hardware
+at runtime.
+
+The crux of this support in libstd is this macro provided by libstd,
+`cfg_feature_enabled!`. The macro will accept one argument, a string literal.
+The string can be any feature passed to `#[target_feature(enable = ...)]` for
+the platform you're compiling for. Finally, the macro will resolve to a `bool`
+result.
+
+For example on x86 you could write:
+
+```rust
+if cfg_feature_enabled!("sse4.1") {
+    println!("this cpu has sse4.1 features enabled!");
+}
+```
+
+It would, however, be an error to write this on x86 cpus:
+
+```rust
+cfg_feature_enabled!("neon"); //~ ERROR: neon is an ARM feature, not x86
+cfg_feature_enabled!("foo"); //~ ERROR: unknown target feature for x86
+```
+
+The macro is intended to be implemented in the `std` crate (**not** `core`) and
+made available via the normal macro preludes. The implementation of this macro
+is expected to be what [`stdsimd`][stdsimd] does today, notably:
+
+* The first time the macro is invoked all the local CPU features will be
+  detected.
+* The detected features will then be cached globally (currently in a bitset) for
+  the rest of the execution of the program.
+* Further invocations of `cfg_feature_enabled!` are expected to be cheap runtime
+  dispatches. (aka load a value and check whether a bit is set)
+* Exception: in some cases the result of the macro is statically known: for
+  example, `cfg_feature_enabled!("sse2")` when the binary is being compiled
+  with "sse42" globally. In these cases, none of the steps above are
+  performed and the macro just expands to `true`.
```

Comment:
So, this seems counterintuitive. Wouldn't it make sense for `cfg_feature_enabled!("neon")` to statically expand to `false` if it is being compiled for `x86`, by the same logic that says `cfg_feature_enabled!("sse2")` should statically expand to `true` when the binary is being compiled for a target that is guaranteed to have it?

Reply:
Making it a compile-time error just seems like it will add a lot of difficulty to using these intrinsics when trying to support multiple architectures, since you'll have to separate out the implementations by arch somehow.

Reply:
> since you'll have to separate out the implementations by arch somehow.

You have to do that anyway? Could you show an example of code you want to write that this would inhibit you from writing?

Reply:
@coder543 for an example like that though this would be statically disallowed (on x86 platforms)

```rust
#[target_feature(enable = "neon")]
fn foo_neon(..) {}
```

Reply:
making `#[target_feature(enable = "neon")]` a compile-time error on x86 makes perfect sense, but I'm really talking about the conditional form which is what was specifically mention on the RFC.

It's entirely possible that I'm just misunderstanding things!

Reply:
@coder543 nah I think I  understand where you're coming from, I'm personally all for ergonomics as well! I think, though, that in this case it ends up not helping much. For example it may be useful taking a look at the [example hex encoder](https://github.com/rust-lang-nursery/stdsimd/blob/master/examples/hex.rs) which currently only has optimizations for x86. Functions like `hex_encode_avx2` *must* have a `#[cfg]` to only exist on x86, so on platforms like ARM that function doesn't exist.

If that's the case then the [portable function](https://github.com/rust-lang-nursery/stdsimd/blob/0d6b868ade601e1159c23418f78902012a246a1f/examples/hex.rs#L42-L59) also *must* have at least a `#[cfg]` for dealing with the x86 functions being called. Put another way, forcing a `#[cfg]` in one location forces a `#[cfg]` in another.

With all that in mind if we changed `cfg_feature_enabled!` to accept all strings (and just expand to `false` for unknown ones) then I don't think we'd actually reduce the amount of `#[cfg]` necessary in the program unfortunately. In that sense I think it's best to stick to a conservative route and only accept whitelisted strings.

Reply:
Ah, I think I see the critical piece I was missing. It makes sense now.

Reply:
Doesn't the macro use to do the runtime check _replace_ the compile-time attribute use?

```rust
if cfg_feature_enabled!("sse4") {
  // Function call or inline code, shouldn't matter.
} else {
  // Fallback code (func or inline)
}
```

And can't the macro expand to a static value in some places? Then why can't requesting something that will clearly never exist on the compile target just be a static false and fold away into always being the else branch fallback code?

Reply:
If I understand correctly, the *intrinsic functions themselves* are behind `cfg` flags, and so they will not even exist *at compile-time* on other architectures, so referencing those non-existent functions would generate an entirely different kind of error, and a more confusing one.

Reply:
@Lokathor it sort of does, yeah, @coder543 is right though in that the target functions you're calling are behind `cfg`, and rustc will still typecheck:

```rust
if false {
    // rustc still typechecks and resolves everything here
}
```

Reply:
Backup strategy: could we have the function outlines exist and then have the bodies be unimplemented!("intrinsic not available on this cpu type"); or some such?

Reply:
@Lokathor perhaps! That'd probably be a separate procedural macro or something like that though.